<a href="https://colab.research.google.com/github/Manekineco1/GLANDA/blob/main/Product_ORMC%2C_pytesseract%2C_cv2_%22%22.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Разработка решения по распознаванию донорских справок для Donorsearch, команда 10, "Это для справки"

# Написание функций распознавания и загрузки документов

Импортируем все необходимые библиотеки

In [8]:
%%capture
!sudo apt install tesseract-ocr
!sudo apt-get install tesseract-ocr-rus
!pip install pytesseract

In [9]:
import os, fnmatch
import matplotlib.pyplot as plt
import numpy as np
import pandas as pd
import pytesseract
import cv2
import scipy
from scipy.signal import find_peaks
import re
from google.colab import files

Расшифровываем аббревиатуры.

In [10]:
def vddon(vid):
  if len(vid) == 0:
    return(vid)
  if vid[0].lower() == 'к':
    return 'Цельная кровь'
  if vid[0].lower() == 'п':
    return 'Плазма'
  if vid[0].lower() == 'ц':
    return 'Тромбоциты'
    return('Цельная кровь')

Прописываем функцию для распознавания документов компьютерным зрением.

In [11]:
class ORMC():
    """
    Оптическое распознование медицинских справок
    """
    def __init__(self, namefileimg):
        self.namefileimg = namefileimg
        self.img = cv2.imread(self.namefileimg)

    def scale(self, size = 3000):
      if self.img.shape[0] < size:
        scale_percent = 3000 // self.img.shape[0]
        width = int(self.img.shape[1] * scale_percent)
        height = int(self.img.shape[0] * scale_percent)
        img = cv2.resize(self.img, (width, height),
                         interpolation = cv2.INTER_AREA
                         )
        return self.img

    def predict_raw(self):
      self.text = pytesseract.image_to_string(self.img,
                                              lang='rus',
                                              config='--psm 6'
                                              )
      return self.text


    def predict_data(self):
        ptrn1 = r'([0-9]{2}[.| ][0-9]{2}[.| ][0-9]{4})|([0-9]{8})'
        ptrn2 = '[^0-9-а-яА-Я]+'
        mtch = re.findall(ptrn1, self.text)
        if mtch:
          start = self.text.find( ''.join(mtch[0]))
          end = self.text.rfind( ''.join(mtch[-1]))
          donardates =re.sub(pattern=ptrn2,
                                  repl=' ',
                                  string= self.text[start:end]
                                  )
          rzlt = []
          str1 =  donardates.split(' ')
          for i in range(len(str1)):
            if str1[i].isnumeric() and re.match(r'(201\d|202\d)',str1[i]):
              rzlt.append([f'{str1[i-2]}.{str1[i-1]}.{str1[i]}',
                           vddon(str1[i+1])])
            if str1[i].isnumeric() and len(str1[i])==6:
              rzlt.append([f'{str1[i][:2]}.{str1[i][2:4]}.{str1[i][4:]}',
                           vddon(str1[i+1])])
          self.donardates = pd.DataFrame(rzlt, columns = ['Дата донации', 'Класс крови'])
          self.donardates['Тип донации'] = 'Безвозмездно'
          return  self.donardates
        return None

    def area(self, part = None):
      img = self.img
      if part == 'bottom':
        #нижняя часть дока
        img = img[img.shape[0]//2:img.shape[0], 0:img.shape[1]]
      if part == 'top':
        #верхняя часть дока
        img = img[img.shape[0]//2:img.shape[0], 0:img.shape[1]]
      self.img = img
      return self.img

    def drop_lines(self):
      """удаляет линии таблицы"""
      gray = cv2.cvtColor(self.img, cv2.COLOR_BGR2GRAY)
      thresh = cv2.threshold(gray,
                             0, 255,
                             cv2.THRESH_BINARY_INV + cv2.THRESH_OTSU)[1]
      # Repair horizontal table lines
      kernel = cv2.getStructuringElement(cv2.MORPH_RECT, (2,1))
      thresh = cv2.morphologyEx(thresh, cv2.MORPH_CLOSE, kernel, iterations=1)
      # Remove horizontal lines
      horizontal_kernel = cv2.getStructuringElement(cv2.MORPH_RECT, (55,2))
      detect_horizontal = cv2.morphologyEx(thresh,
                                           cv2.MORPH_OPEN,
                                           horizontal_kernel,
                                           iterations=5
                                           )
      cnts = cv2.findContours(detect_horizontal,
                              cv2.RETR_EXTERNAL,
                              cv2.CHAIN_APPROX_SIMPLE
                              )
      cnts = cnts[0] if len(cnts) == 2 else cnts[1]
      for c in cnts:
          cv2.drawContours(self.img, [c], -1, (255,255,255), 9)
      # Remove vertical lines
      vertical_kernel = cv2.getStructuringElement(cv2.MORPH_RECT, (2,55))
      detect_vertical = cv2.morphologyEx(thresh,
                                         cv2.MORPH_OPEN,
                                         vertical_kernel,
                                         iterations=2)
      cnts = cv2.findContours(detect_vertical,
                              cv2.RETR_EXTERNAL,
                              cv2.CHAIN_APPROX_SIMPLE)
      cnts = cnts[0] if len(cnts) == 2 else cnts[1]
      for c in cnts:
          cv2.drawContours(self.img, [c], -1, (255,255,255), 9)
      return self.img

    def show(self, figsize=(10,14)):
      """показывает изображение сканированого документа """
      plt.figure(figsize=figsize)
      plt.imshow(self.img)
      plt.axis('off')
      plt.show()

    def fit(self):
      self.area('bottom'),
      self.drop_lines(),
      self.predict_raw(),
      df = self.predict_data()
      return df

Прописываем функцию загрузки файлов

In [12]:
def load():
  uploaded = files.upload()
  filename = next(iter(uploaded))
  return filename

Выводим результат загрузки

In [13]:
if __name__ == "__main__":
  filename = load()
  img = ORMC(filename)
  rezult = img.fit()
  print(rezult)

Saving 141899 .jpg to 141899 .jpg
Saving 204119 .jpg to 204119 .jpg
Saving 213950.jpg to 213950.jpg
Saving 225629 .jpg to 225629 .jpg
Saving 227414.jpg to 227414.jpg
Saving 228963 .jpg to 228963 .jpg
Saving 231820 .jpg to 231820 .jpg
Saving 233749 .jpg to 233749 .jpg
Saving 236000 .jpg to 236000 .jpg
Saving 238716.jpg to 238716.jpg
Saving 240493.jpg to 240493.jpg
Saving 243478 .jpg to 243478 .jpg
Saving 245365 .jpg to 245365 .jpg
Saving 254586 .jpg to 254586 .jpg
Saving 256578 .jpg to 256578 .jpg
Saving 256838.jpg to 256838.jpg
  Дата донации Класс крови   Тип донации
0     50.50.55      Плазма  Безвозмездно
1     42.20.22        None  Безвозмездно
2     90.06.02      Плазма  Безвозмездно
3     19.07.20        None  Безвозмездно


Промежуточный вывод:
- Нам в работу переданы 16 файлов в формате jpg, в которых нам нужно распознать данные в трех колонках - 'Дата донации', 'Класс крови',  'Тип донации'.
- Нами была написана функция по распознаванию донорских справок в формате jpg с помощью библиотеки pytesseract
- Мы загрузили необходимые файлы через гугл-диск



##Проверка

Подключаем гугл-диск

In [14]:
from google.colab import drive
drive.mount('/content/drive', force_remount=True)

Mounted at /content/drive


Указываем путь к файлам

In [15]:
PATH = '/content/drive/MyDrive/Colab Notebooks/datasets/405/405'
content = os.listdir(PATH)

Указываем, какие файлы подтягивать из директоии

In [16]:
pattern1 = r'([0-9]{2}[.| ][0-9]{2}[.| ][0-9]{4})|([0-9]{8})'
imspv = []
pattern = "*.jpg"
for entry in content:
    if fnmatch.fnmatch(entry, pattern):
            imspv.append(entry)
imspv

['238716.jpg',
 '245365 .jpg',
 '233749 .jpg',
 '204119 .jpg',
 '254586 .jpg',
 '231820 .jpg',
 '236000 .jpg',
 '225629 .jpg',
 '228963 .jpg',
 '141899 .jpg',
 '243478 .jpg',
 '227414.jpg',
 '213950.jpg',
 '256578 .jpg',
 '256838.jpg']

Применяем метод распознавания к списку файлов и создаем новый экземпляр класса.

In [17]:
rzltdf = {}
for f in imspv:
  img = ORMC(f'{PATH}/{f}')
  rzltdf[f[:6]] = img.fit()

Выводим результат распознавания

In [18]:
for img in rzltdf:
  print()
  print("\033[1m" +img+ "\033[0m")
  print(rzltdf[img])
  print('#'*50)


238716
   Дата донации    Класс крови   Тип донации
0    10.08.2017  Цельная кровь  Безвозмездно
1    20.10.2020  Цельная кровь  Безвозмездно
2    05.10.2013  Цельная кровь  Безвозмездно
3    16.12.2017  Цельная кровь  Безвозмездно
4    23.12.2020  Цельная кровь  Безвозмездно
5    27.03.2014  Цельная кровь  Безвозмездно
6    26.04.2018  Цельная кровь  Безвозмездно
7    17.02.2021         Плазма  Безвозмездно
8    29.07.2015  Цельная кровь  Безвозмездно
9    12.07.2018  Цельная кровь  Безвозмездно
10   03.03.2021  Цельная кровь  Безвозмездно
11   03.10.2015  Цельная кровь  Безвозмездно
12   15.11.2018  Цельная кровь  Безвозмездно
13   04.05.2021  Цельная кровь  Безвозмездно
14   18.02.2016  Цельная кровь  Безвозмездно
15   27.03.2019  Цельная кровь  Безвозмездно
16   04.08.2021  Цельная кровь  Безвозмездно
17   19.04.2016  Цельная кровь  Безвозмездно
18   06.06.2019  Цельная кровь  Безвозмездно
19   29.10.2021  Цельная кровь  Безвозмездно
20   07.07.2016  Цельная кровь  Безвозмездно
21

Подтягиваем CSV файлы с разметкой из директории

In [22]:
csvspv = []
import os, fnmatch
content = os.listdir(PATH)
pattern = "*.csv"
for entry in content:
    if fnmatch.fnmatch(entry, pattern):
            csvspv.append(entry)
csvspv

['227414.csv',
 '231820.csv',
 '233749.csv',
 '228963.csv',
 '225629.csv',
 '238716.csv',
 '141899.csv',
 '213950.csv',
 '204119.csv',
 '256838.csv',
 '256578.csv',
 '245365.csv',
 '243478.csv',
 '254586.csv',
 '236000.csv']

Читаем CSV файлы

In [23]:
csvdf = {}
for csv in csvspv:
  csvdf[csv[:6]]=pd.read_csv(f'{PATH}/{csv}')

Пишем функцию метрики accuracy, которая сравнивает содержимое ячеек в трех столбцах - 'Дата донации', 'Класс крови', 'Тип донации' в распознанных файлах и в файлах с разметкой . Выводим метрику для каждого файла и общую.

In [24]:
def mtrc():
  mtrc = []
  number_coinciden = 0
  total_elements = 0
  for csv in csvdf:
    print()
    print("\033[1m" +csv+ "\033[0m")
    if csv in rzltdf.keys() and rzltdf[csv] is not None:
      test = set(csvdf[csv]['Дата донации'] +' '+ csvdf[csv]['Класс крови'] + ' ' + csvdf[csv]['Тип донации'] )
      predict = set(rzltdf[csv]['Дата донации'] +' '+ rzltdf[csv]['Класс крови'] + ' ' + rzltdf[csv]['Тип донации'] )
      total_elements +=len(test)
      number_coinciden += len(test& predict)
      mtrk = len(test& predict) / len(test)
      mtrc.append(mtrk)
      print(round(mtrk,4))

  print()
  print('Общая Метрика')
  return round(number_coinciden/total_elements,4)
mtrc()


227414

231820
0.7692

233749
0.381

228963

225629
0.7222

238716
0.9444

141899
0.0

213950
0.6818

204119
0.6667

256838
0.0455

256578

245365
0.8378

243478

254586
0.0

236000
0.4

Общая Метрика


0.5364

Общий вывод:
- Нами была проведена работа по распознаванию донорских справок в формате jpg  с помощью библиотеки pytesseract
- Написана функция метрики accuracy, которая сравнивает содержимое ячеек в трех столбцах - 'Дата донации', 'Класс крови', 'Тип донации'.
- Выведена метрика для каждого файла и общая.
- Общая accuracy составила 0.5364.
- В силу того, что в нашей команде изначально было три человека, но один из участников не работал над проектом, и потом ему искали замену и только 5 июля к нам присоединился третий участник - мы не успели отладить методы предобработки изображений, поэтому убрали соотвествующий код. А так же мы не успели обернуть проект в микросервис. Просим дать нам время до 21.07.23. За это время мы сможем улучшить метрику и скорее всего успеем обернуть в микросервис.